File: src/main/java/com/example/tutorial/dsa/medium/concurrency/producerconsumer/BoundedBuffer.java
package com.example.tutorial.dsa.medium.concurrency.producerconsumer;

import com.example.tutorial.common.datamodel.Task;

import java.util.LinkedList;
import java.util.Queue;

/**
 * BoundedBuffer implementing Producer-Consumer pattern with wait/notify.
 *
 * <p><b>LOGIC (wait/notify Synchronization):</b>
 * <ol>
 *   <li>Producers wait when buffer is full</li>
 *   <li>Consumers wait when buffer is empty</li>
 *   <li>notifyAll() wakes up waiting threads when state changes</li>
 * </ol>
 *
 * <p><b>How it works:</b>
 * <pre>
 * Buffer capacity: 3
 *
 * Producer tries to add when full:
 *   → Calls wait(), releases lock, sleeps
 *   → Consumer removes item, calls notifyAll()
 *   → Producer wakes up, re-acquires lock, adds item
 *
 * Consumer tries to take when empty:
 *   → Calls wait(), releases lock, sleeps
 *   → Producer adds item, calls notifyAll()
 *   → Consumer wakes up, re-acquires lock, takes item
 * </pre>
 *
 * <p><b>Why while loop instead of if?</b>
 * <pre>
 * // WRONG - spurious wakeup can cause issues
 * if (buffer.size() == capacity) {
 *   wait();
 * }
 *
 * // CORRECT - re-check condition after wakeup
 * while (buffer.size() == capacity) {
 *   wait();
 * }
 * </pre>
 *
 * <p><b>Time Complexity: O(1)</b> for put and take (excluding wait time).
 * <p><b>Space Complexity: O(capacity)</b>
 *
 * <p><b>When to use:</b> Educational purposes. In production, prefer
 * {@link java.util.concurrent.BlockingQueue} implementations.
 *
 * @param <T> type of items in the buffer
 */
public class BoundedBuffer<T> {
  private final Queue<T> buffer;
  private final int capacity;

  public BoundedBuffer(int capacity) {
    this.capacity = capacity;
    this.buffer = new LinkedList<>();
  }

  /**
   * Adds an item to the buffer. Blocks if buffer is full.
   *
   * @param item the item to add
   * @throws InterruptedException if thread is interrupted while waiting
   */
  public synchronized void put(T item) throws InterruptedException {
    while (buffer.size() == capacity) {
      wait();  // Buffer full, wait for consumer.
    }
    buffer.add(item);
    notifyAll();  // Wake up consumers.
  }

  /**
   * Takes an item from the buffer. Blocks if buffer is empty.
   *
   * @return the item taken
   * @throws InterruptedException if thread is interrupted while waiting
   */
  public synchronized T take() throws InterruptedException {
    while (buffer.isEmpty()) {
      wait();  // Buffer empty, wait for producer.
    }
    T item = buffer.poll();
    notifyAll();  // Wake up producers.
    return item;
  }

  /**
   * Returns current buffer size.
   */
  public synchronized int size() {
    return buffer.size();
  }

  /**
   * Checks if buffer is empty.
   */
  public synchronized boolean isEmpty() {
    return buffer.isEmpty();
  }

  /**
   * Checks if buffer is full.
   */
  public synchronized boolean isFull() {
    return buffer.size() == capacity;
  }
}

File: src/main/python/com/example/tutorial/dsa/medium/concurrency/producerconsumer/bounded_buffer.py
"""
BoundedBuffer implementing Producer-Consumer pattern with Condition variables.

LOGIC (wait/notify Synchronization):
    1. Producers wait when buffer is full
    2. Consumers wait when buffer is empty
    3. notify_all() wakes up waiting threads when state changes

How it works:
    Buffer capacity: 3

    Producer tries to add when full:
      -> Calls wait(), releases lock, sleeps
      -> Consumer removes item, calls notify_all()
      -> Producer wakes up, re-acquires lock, adds item

    Consumer tries to take when empty:
      -> Calls wait(), releases lock, sleeps
      -> Producer adds item, calls notify_all()
      -> Consumer wakes up, re-acquires lock, takes item

Time Complexity: O(1) for put and take (excluding wait time).
Space Complexity: O(capacity)

When to use: Educational purposes. In production, prefer queue.Queue.
"""

import threading
from collections import deque
from typing import TypeVar, Generic

T = TypeVar('T')


class BoundedBuffer(Generic[T]):
    """Thread-safe bounded buffer for producer-consumer pattern."""

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.buffer = deque()
        self.lock = threading.Lock()
        self.not_full = threading.Condition(self.lock)
        self.not_empty = threading.Condition(self.lock)

    def put(self, item: T) -> None:
        """
        Adds an item to the buffer. Blocks if buffer is full.

        Args:
            item: The item to add
        """
        with self.not_full:
            while len(self.buffer) >= self.capacity:
                self.not_full.wait()  # Buffer full, wait for consumer.
            self.buffer.append(item)
            self.not_empty.notify_all()  # Wake up consumers.

    def take(self) -> T:
        """
        Takes an item from the buffer. Blocks if buffer is empty.

        Returns:
            The item taken from the buffer
        """
        with self.not_empty:
            while len(self.buffer) == 0:
                self.not_empty.wait()  # Buffer empty, wait for producer.
            item = self.buffer.popleft()
            self.not_full.notify_all()  # Wake up producers.
            return item

    def size(self) -> int:
        """Returns current buffer size."""
        with self.lock:
            return len(self.buffer)

    def is_empty(self) -> bool:
        """Checks if buffer is empty."""
        with self.lock:
            return len(self.buffer) == 0

    def is_full(self) -> bool:
        """Checks if buffer is full."""
        with self.lock:
            return len(self.buffer) >= self.capacity
